### DISCUSSION TITLE
Diagnosing "configuration uses unsupported authentication" to greater detail

### DISCUSSION THREAD

**Discussion by bernd-wechner** [strongswan/strongswan#2326] [Q&A]
I'm trying to connect to a work VPN, that's just gone from SSL to IPSEC. They've provided support for Windows and Mac empoylees and I'm the sore thumb with Linux at home and it's not working.  Work side is a Fotigate firewall and they provide a Linux FortiClient but:

- it's not working and provides no lucid diagnostics beyond authentication errors
- it's a GUI client and I'd much rather use the CLI to raise the connection.

To wit, experimenting with strongswan as the net is full of info and advice and I harbour the hope it can provide better diagnostics. 

Alas, on configuring it and trying to bring the connection up, I get a basic failure reporting "configuration uses unsupported authentication".

My basic question is how I can drill down to get details on the exact meaning of that. I might infer that server and client side (left and right I guess) aren't agreed on the authentication means. So to clarify, I'm in a Linux Mint box (Ubuntu base) and am informed that we are using IKEv1 (apparently only because the Mac users can't get IKEv2 working aaargh) with a certificate and a username/password (xauth) as a sort of 2FA level.

They've provided me a certificate and because it's self-signed, a ca certificate and a private key.

I can of course provide all my configs, and I have indeed done a good deal of reading. But for now, I'm really asking a more basic question that I've found no answer to: Is there a way to expand on this failure. That is, inform me of what the precise point of failure is. Under this one message might fall all these:

1. A problem with the cert
2. A problem with the cacert
3. A problem with the private key
4. A problem with the username
5. A problem with the password (not specified anywhere at present)
6. A problem with the IKE version

Essentially, I'm curious if we can drill down to what exactly is failing?

I target of course upskilling, learning not just a config that works, but how to diagnose and fix configs ideally.

As a backdrop and facilitate an informed response if it's on offer the configs (anonymised):

```
$ sudo tree -f /etc/ipsec.d
/etc/ipsec.d
├── /etc/ipsec.d/aacerts
│   └── /etc/ipsec.d/aacerts/my-cert.pem
├── /etc/ipsec.d/acerts
├── /etc/ipsec.d/cacerts
│   └── /etc/ipsec.d/cacerts/root-cert.pem
├── /etc/ipsec.d/certs
├── /etc/ipsec.d/crls
├── /etc/ipsec.d/ocspcerts
├── /etc/ipsec.d/policies
├── /etc/ipsec.d/private
│   └── /etc/ipsec.d/private/my-key.pem
└── /etc/ipsec.d/reqs

9 directories, 3 files

$ sudo cat /etc/ipsec.secrets
# This file holds shared secrets or RSA private keys for authentication.

# RSA private key for this host, authenticating it to any other host
# which knows the public part.

: RSA /etc/ipsec.d/private/my-key.pem

$ sudo cat /etc/ipsec.conf
# ipsec.conf - strongSwan IPsec configuration file
conn %default
    keyexchange=ikev1
    authby=rsasig
    ike=aes256-sha1-modp1024
    esp=aes256-sha1-modp1024  # Enable PFS by including a DH group here

conn work
    aggressive=yes
    ikelifetime=60m
    keylife=20m
    rekeymargin=3m
    keyingtries=1
    left=%defaultroute
    leftcert=/etc/ipsec.d/aacerts/my-cert.pem
    leftfirewall=yes
    leftsourceip=%config
    right=fw.work.com
    rightsubnet=192.168.1.54/32 # Only route traffic to this specific IP
    rightid=@berndw
    rightca=/etc/ipsec.d/cacerts/root-cert.pem
    rightauth=eap-mschapv2
    eap_identity=my_username
    auto=add
    authby=rsasig

$ sudo ipsec start
Starting strongSwan 5.9.5 IPsec [starter]...
$ sudo ipsec up work
initiating Aggressive Mode IKE_SA work[1] to 120.151.44.39
configuration uses unsupported authentication
establishing connection 'work' failed
```
The ipsec.conf I admit fully I have based on on-line docs and chatgpt's suggestions and a todo is to read up on each of these options and work out what they precisely do, but short term I'd like to have a better diagnostic loop and work out the point and cause of failure. Given my skill level and the rapid fire guess work and research thus far it could well be something simple like eap-mschapv2 not being the right setting for rightauth.

I can also provide these screenshots of the Fortigate side server configuration that I was provided with:

![image](https://github.com/strongswan/strongswan/assets/7296506/b23164b8-1ba3-4a21-86ff-3b48ad5c33d9)

![image](https://github.com/strongswan/strongswan/assets/7296506/8bd427da-f2c3-45ac-9bec-f4548905c3ff)

I'm hoping that I've just overlooked something simple in the way of extracting deeper diagnostics. And apologise that I've not made/had time to read up on all those many config settings in detail yet, though I have read many of the recommended intro pages.

**Comment by tobiasbrunner** [comment]
> Given my skill level and the rapid fire guess work and research thus far it could well be something simple like eap-mschapv2 not being the right setting for rightauth.

Correct, EAP is only available for IKEv2.

Since you received a client certificate and private key, why would you assume the authentication is something different than `pubkey`? Or do you also have a username/password?

**Reply by bernd-wechner** [reply]
Now I just have to work out the right settings ... as time permits I'll get closer to finding deeper debug traces I hope. Posted here in the hope of some pointers. There is so so much to read ...

**Comment by bernd-wechner** [comment]
Well I have file logging and debugging working at last and have walked through all the configuration settings. Key learnings so far I think are:

1. It's complicated ;-). Lots to read and learn.
2. There are two key conf files `/etc/ipsec.conf` and `/etc/strongswan.conf` which are both apparently read and supported. I'm on an Unbuntu derivative and so have the strongswan package installed and use `sudo ipsec start` to bring the daemon up, and `ipsec up myconnection` to start my connection.
3. Logging can be configured in `ipsec.conf` but apparently only the levels not the targets but `strongswan.conf` comes packaged with templates in `strongswan.d/charon-logging.conf` and I configured that and am now seeing /var/log/charon.log. Caveat: apparmor on Ubuntu prevented writing the log file and a line was needed in `/etc/apparmor.d/usr.lib.ipsec.charon` (`/var/log/charon.log       rw,` does the trick) and then logging works fine. 
4. I have a feeling I should remove the `ipsec.conf` connection configuration and move entirely to `strongswan.conf`, because of: https://docs.strongswan.org/docs/6.0/config/config.html. Certainly having two paths for configuration is confusing and ipsec.conf seems deprecated.

Which now sees me with good logging anyhow, I've fixed the settings above to get a much better connection effort but am stil failing but that is a different thread. This one was to see if anyone could fast track me to logging configs that make sense on Ubuntu (whcih didn't happen but that's fine, we're a community not a paid support system ;-). 

In summary on Unbuntu 

1. Install strongswan: `sudo apt install strongswann`
2. Confirm that `/etc/strongswan.conf` includes `/etc/strongswan.d/charon-logging.conf/` and that `/etc/strongswan.d/charon-logging.conf/` exists
3. Configure your logging. Here's what I went with in `/etc/strongswan.d/charon-logging.conf/`:
    ```
    charon {
        # Section to define file loggers, see LOGGER CONFIGURATION in
        # strongswan.conf(5).
        filelog {
    
            # <name> may be the full path to the log file if it only contains
            # characters permitted in section names. Is ignored if path is
            # specified.
	    # 
	    # Since version 5.7.0 section names can’t contain dots or colons. 
	    # To specify such paths use the path setting in an arbitrarily 
	    # named section.
            default {
    
                # Loglevel for a specific subsystem.
                # <subsystem> = <default>
                cfg = 2  # Configuration management and plugins
	        ike = 4  # IKE_SA/ISAKMP SA
	        net = 1  # IKE network communication
                mgr = 4  # IKE_SA manager, handling synchronization for IKE_SA access
                knl = 2  # IPsec/Networking kernel interface
                esp = 4  # libipsec library messages
                dmn = 4  # Main daemon setup/cleanup/signal handling
    
                # If this option is enabled log entries are appended to the existing file.
                append = yes
    
                # Default loglevel.
                default = 1
    
                # Enabling this option disables block buffering and enables line
                # buffering.
                flush_line = yes
    
                # Prefix each log entry with the connection name and a unique
                # numerical identifier for each IKE_SA.
                ike_name = yes
    
                # Add the log level of each message after the subsystem (e.g. [IKE2]).
                log_level = yes
    
                # Optional path to the log file. Overrides the section name. Must be
                # used if the path contains characters that aren't allowed in
                # section names.
                path = /var/log/charon.log
    
                # Adds the milliseconds within the current second after the
                # timestamp (separated by a dot, so time_format should end with %S
                # or %T).
                time_add_ms = yes
    
                # Prefix each log entry with a timestamp. The option accepts a
                # format string as passed to strftime(3).
                time_format = %F %T
            }
    
        }
    }    ```
4. Configure apparmor (which is installed on Ubuntu by default and the strongswan package configured with it, to provide a layer of security, it will prevent strongswan from writing your log file unless you permit it with apparmor). In `/etc/apparmor.d/usr.lib.ipsec.charon` add a line permitting it. I went with `/var/log/charon.log       rw,`.
5. Restart the daemon `sudo ipsec restart`
6. Check the log: `cat /var/log/charon.log`
7. Try to bring your conenction up and check the log. Now the logs are isolated and not mixed int with syslog and authlog and such and you can tweak the log levels and make some progress in diagnosing your connection.

**Reply by tobiasbrunner** [reply]
>     2. There are two key conf files `/etc/ipsec.conf` and `/etc/strongswan.conf` which are both apparently read and supported. I'm on an Unbuntu derivative and so have the strongswan package installed and use `sudo ipsec start` to bring the daemon up, and `ipsec up myconnection` to start my connection.

Note that `ipsec.conf` and `ipsec` is deprecated. The recommended backend is [swanctl.conf](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html)/[swanctl](https://docs.strongswan.org/docs/5.9/swanctl/swanctl.html). On Debian/Ubuntu this is provided by a separate package (`strongswan-swanctl`). There are also two IKE daemons (starter/charon and charon-systemd) of which only one should be installed (or enabled in systemd).

>     3. Logging can be configured in `ipsec.conf` but apparently only the levels not the targets but `strongswan.conf` comes packaged with templates in `strongswan.d/charon-logging.conf` and I configured that and am now seeing /var/log/charon.log. Caveat: apparmor on Ubuntu prevented writing the log file and a line was needed in `/etc/apparmor.d/usr.lib.ipsec.charon` (`/var/log/charon.log       rw,` does the trick) and then logging works fine.

Yes, see [here](https://docs.strongswan.org/docs/5.9/config/logging.html) for details.

>     4. I have a feeling I should remove the `ipsec.conf` connection configuration and move entirely to `strongswan.conf`, because of: https://docs.strongswan.org/docs/6.0/config/config.html. Certainly having two paths for configuration is confusing and ipsec.conf seems deprecated.


No, [strongswan.conf](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html) contains global settings, no connections or credentials. The latter is handled by swanctl.conf.

> In summary on Unbuntu
> 
>     1. Install strongswan: `sudo apt install strongswann`

This meta package currently installs the legacy configuration backend. It's fine to use it still, but definitely not recommended.

>     2. Confirm that `/etc/strongswan.conf` includes `/etc/strongswan.d/charon-logging.conf/` and that `/etc/strongswan.d/charon-logging.conf/` exists

Sure, although, you can also just edit `/etc/strongswan.conf` and put your settings there. Note that the latter should not be a directory but a file (i.e. no slash at the end I guess).

>     5. Restart the daemon `sudo ipsec restart`

While you could do that I'd recommend to use `systemctl` even for the legacy daemon (i.e. `systemctl restart strongswan-starter`)
 
>     6. Check the log: `cat /var/log/charon.log`
> 
>     7. Try to bring your conenction up and check the log. Now the logs are isolated and not mixed int with syslog and authlog and such and you can tweak the log levels and make some progress in diagnosing your connection.

With `journalctl` you can use `-u strongswan-starter.service` (or `-u strongswan.service` when using the modern daemon) to limit the log to just the IKE daemon's output.

**Reply by bernd-wechner** [reply]
Thanks enormously Tobias. 

A couple of questions to speed me up if possible:

> There are also two IKE daemons (starter/charon and charon-systemd) of which only one should be installed (or enabled in systemd).

1. How can I tell if I've got starter/charon or charon-systemd or both and if I have only the former how do I get the latter on an ubuntu system and ensure the former is removed (disabled). Sadly I didn't find much useful info in all the docs I've read thus far on that front.
2. I shall install strongswan-swanctl and migrate to that (future proofing - I was following online guides and I guess they are all dated which is the way at times with the internet when there are breaking changes in software or changes in paradigm)
3. Re: logging, I prefer a dedicated log over journalctl myself but that's not a biggie. Currently there's nothing in journalctl because the systemd service is starter and doesn't stay running ...

We have thus far divined, from the fortigate logs that my connection is being rejected after charon sends my personal cert (it's flagged as invalid server side). It seems fortigate has a passive failure mode as charon continues thinking all is good but from that point on gets no further response from the remote end it seems.

**Reply by tobiasbrunner** [reply]
> 1. How can I tell if I've got starter/charon or charon-systemd or both and if I have only the former how do I get the latter on an ubuntu system and ensure the former is removed (disabled). Sadly I didn't find much useful info in all the docs I've read thus far on that front.

The `strongswan-starter` and `strongswan-charon` packages provide the legacy daemon(s) (in particular the former provides the `strongswan-starter` systemd service. The `charon-systemd` package provides the daemon of the same name and the `strongswan` service. So make sure you have only the package(s) installed you want.



> 2\. I shall install strongswan-swanctl and migrate to that (future proofing - I was following online guides and I guess they are all dated which is the way at times with the internet when there are breaking changes in software or changes in paradigm)

The `vici` plugin and `swanctl` tool can also be used with the old daemon when this package is installed. However, only the `strongswan` systemd service loads the config automatically (i.e. calls `swanctl --load-all`).



> 3\. Re: logging, I prefer a dedicated log over journalctl myself but that's not a biggie. Currently there's nothing in journalctl because the systemd service is starter and doesn't stay running ...

It should stay running as it's kind of like a watchdog daemon that controls charon. But yeah, possible that logging from the latter via syslog -> journald ends up in some other bucket and is not associated with the `strongswan-starter` unit.

> We have thus far divined, from the fortigate logs that my connection is being rejected after charon sends my personal cert (it's flagged as invalid server side).

Not really what we saw in the client log (although that was only a small excerpt). But you definitely have to fix that if the server complains. Also, does the Fortigate box not support IKEv2?

**Reply by bernd-wechner** [reply]
Well, I solved the cert problem and now have xauth problem but am proceeding. I did remove `rightauth2=xauth` but no luck. 

I am however moving away from the deprecated system and started anew (backing up my notes and config to date).

To wit, I uninstalled and purged it all and installed `charon-systemd` and `strongswan-swanctl` (hardly an intuitive config, and I await the day Ubuntu support for this is simply all in the `strongswan` package ;-).

Alas, I'm torn with mild confusion again. As there is `strongswan.conf` and there is `swantctl/swantctl.conf` and no easily found clarity on the roles these play. My tentative reading (and I have read a fair bit) is that:

1. `ipsec.conf` is the deprecated system
2. `strongswan.conf` is the supported current system
3. `swanctl/swanctl.conf`  is the even newer even better system that manages `strongswan.conf` for you ... and has the same internal syntax.

Not 100% sure on this, still reading, but updating as I do seriously appreciate any tips from the experienced and with timezones and all firing off an update seems sensible. I'm getting there. I think once I have the IKEv1 handshake complete cert and xauth, I'll be rolling. 

Oh, and re: IKEv2, alas not supported. The reason is bizarre but also I can't change it. The reason is that our helpdesk supports Windows and Mac clients, and Macs apparently do not support IKEv2! So they are only offering IKEv1 for now. To double the irony, the FortiClient they recommend, they tell me on Linux only supports IKEv2. Hence I'm here with Strongswan, which not only promises to work, but unlike the FortiClient has rich configs and extremely detailed debug logging to support configuration success which the FortiClient lacks for the most part.

**Reply by bernd-wechner** [reply]
Dang, the change to swanctl is hard.  I was getting the IKEv1 handshake past the cert and to the xauth stage and failing there with ipsec.conf, but now I'm back to:

2024-07-12 09:28:15.299 05[CFG1] <Work_VPN|2> configuration uses unsupported authentication

which I got past ages ago in ipsec.conf. The solution in ipsec.conf was to specify:

    ike=aes256-sha256-modp2048

But if I use in my swanctl.conf

        proposals = aes256-sha256-modp2048
 
 Apparently that's not the equivalent More reading to do. Alas log levls high and all it says is:
 
2024-07-12 09:28:15.299 05[IKE2] <Work_VPN|2> IKE_SA Work_VPN[2] state change: CREATED => CONNECTING
2024-07-12 09:28:15.299 05[CFG1] <Work_VPN|2> configuration uses unsupported authentication
2024-07-12 09:28:15.299 05[MGR2] <Work_VPN|2> checkin and destroy IKE_SA Work_VPN[2]
2024-07-12 09:28:15.299 05[IKE2] <Work_VPN|2> IKE_SA Work_VPN[2] state change: CONNECTING => DESTROYING

Even if I up logging to level4 default (on all modules, all I see is:

2024-07-12 09:33:34.034 15[IKE2] <Work_VPN|1> IKE_SA Work_VPN[1] state change: CREATED => CONNECTING
2024-07-12 09:33:34.034 02[JOB3]   watching 10 for reading
2024-07-12 09:33:34.034 15[CFG1] <Work_VPN|1> configuration uses unsupported authentication

Be nice if it logged the proposal negotiation ....

**Reply by bernd-wechner** [reply]
No luck so far. I have tied proposals:

        proposals = "aes128-sha1-modp1536,aes128-sha1-modp2048,aes128-sha256-modp1536,aes128-sha256-modp2048,aes256-sha1-modp1536,aes256-sha1-modp2048,aes256-sha256-modp1536,aes256-sha256-modp2048"

which covers all those supported (as shared in OP). I find the docs rather uncertain. This page:

https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html

under proposals only says: 

Use multiple proposals to offer different algorithm combinations with IKEv1. Algorithm keywords get separated using dashes. Multiple proposals may be separated by commas. 

without providing any useful examples alas. And from:

https://wiki.strongswan.org/projects/strongswan/wiki/IKEv1CipherSuites

I see the algorithm keywords, but again no clear examples, but it does state I can use these in swanctl.conf. 

And it look slike I have everything right but still no go. A full debug log:

2024-07-12 10:04:30.196 07[CFG2] vici client 2 requests: initiate
2024-07-12 10:04:30.196 07[CFG1] vici initiate IKE_SA 'Work_VPN'
2024-07-12 10:04:30.196 09[MGR2] checkout IKE_SA by config
2024-07-12 10:04:30.196 09[MGR2] created IKE_SA (unnamed)[1]
2024-07-12 10:04:30.197 17[LIB2] created thread 17 [591796]
2024-07-12 10:04:30.198 09[KNL2] <Work_VPN|1> using 192.168.0.11 as address to reach 101.178.197.207/32
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> queueing ISAKMP_VENDOR task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> queueing ISAKMP_CERT_PRE task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> queueing AGGRESSIVE_MODE task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> queueing ISAKMP_CERT_POST task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> queueing ISAKMP_NATD task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> activating new tasks
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1>   activating ISAKMP_VENDOR task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1>   activating ISAKMP_CERT_PRE task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1>   activating AGGRESSIVE_MODE task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1>   activating ISAKMP_CERT_POST task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1>   activating ISAKMP_NATD task
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> sending XAuth vendor ID
2024-07-12 10:04:30.198 09[ENC2] <Work_VPN|1> added payload of type VENDOR_ID_V1 to message
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> sending DPD vendor ID
2024-07-12 10:04:30.198 09[ENC2] <Work_VPN|1> added payload of type VENDOR_ID_V1 to message
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> sending FRAGMENTATION vendor ID
2024-07-12 10:04:30.198 09[ENC2] <Work_VPN|1> added payload of type VENDOR_ID_V1 to message
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> sending NAT-T (RFC 3947) vendor ID
2024-07-12 10:04:30.198 09[ENC2] <Work_VPN|1> added payload of type VENDOR_ID_V1 to message
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> sending draft-ietf-ipsec-nat-t-ike-02\n vendor ID
2024-07-12 10:04:30.198 09[ENC2] <Work_VPN|1> added payload of type VENDOR_ID_V1 to message
2024-07-12 10:04:30.198 09[IKE0] <Work_VPN|1> initiating Aggressive Mode IKE_SA Work_VPN[1] to 101.178.197.207
2024-07-12 10:04:30.198 09[IKE2] <Work_VPN|1> IKE_SA Work_VPN[1] state change: CREATED => CONNECTING
2024-07-12 10:04:30.198 02[JOB2] watcher got notification, rebuilding
2024-07-12 10:04:30.199 02[JOB3]   watching 10 for reading
2024-07-12 10:04:30.199 09[CFG1] <Work_VPN|1> configuration uses unsupported authentication

**Reply by bernd-wechner** [reply]
Aargh, I found a clue here under Host to Host:

https://docs.strongswan.org/docs/5.9/config/quickstart.html

and tried adding a remote block and now it works. Well proceeds anyhow. I figured if I was happy to accept the remote side unconditionally I'd not need one but no go. I have to add its pubkey ID to get past.

**Reply by bernd-wechner** [reply]
And I have success!

initiate completed successfully

**Reply by bernd-wechner** [reply]
Alas at the server side (VPN gateway) phase 1 is authentication is complete but not phase 2, but the local side says the connection is initiated.  Helpdesk tells me this requires PFS. More reading ...

**Reply by bernd-wechner** [reply]
Alas Phase 2 is failing. In response to a INFORMATIONAL_V1 request it receives DELETE as a response. If only I understood why.

**Reply by tobiasbrunner** [reply]
> * `ipsec.conf` is the deprecated system

Correct.

> * `strongswan.conf` is the supported current system

No, this is for general settings, no matter what configuration backend is used.

> * `swanctl/swanctl.conf`  is the even newer even better system that manages `strongswan.conf` for you ... and has the same internal syntax.

This is the new configuration/credential config backend. It has nothing directly to do with strongswan.conf other than the shared syntax.

There is an [overview in the docs](https://docs.strongswan.org/docs/5.9/config/config.html).


> The reason is that our helpdesk supports Windows and Mac clients, and Macs apparently do not support IKEv2!

Hehe, OK. While Windows 7 was the first commercial OS that supported IKEv2 out of the box 15 years ago, Apple clients (iOS and macOS) also support IKEv2 since about 9 years. So that really shouldn't be a reason.

**Comment by bernd-wechner** [comment]
I'm happy the original question is resolved (logging in detail is now possible). Closing this and migrating residual issue to:

https://github.com/strongswan/strongswan/discussions/2343
